src/Domain/BlockLogicalCoordinates.cpp

```diff
@@ -0,0 +1,128 @@
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "Domain/Domain.hpp"  // IWYU pragma: keep
+#include "ErrorHandling/Error.hpp"
+#include "Utilities/GenerateInstantiations.hpp"
+
+template <size_t Dim>
+void BlockLogicalCoordHolder<Dim>::pup(PUP::er& p) noexcept {
+  p | block_id;
+  p | block_logical_coords;
+}
+
+template <size_t Dim>
+bool operator==(const BlockLogicalCoordHolder<Dim>& lhs,
+                const BlockLogicalCoordHolder<Dim>& rhs) noexcept {
+  return lhs.block_id == rhs.block_id and
+         lhs.block_logical_coords == rhs.block_logical_coords;
+}
+template <size_t Dim>
+bool operator!=(const BlockLogicalCoordHolder<Dim>& lhs,
+                const BlockLogicalCoordHolder<Dim>& rhs) noexcept {
+  return not(lhs == rhs);
+}
+template <size_t Dim>
+std::ostream& operator<<(std::ostream& os,
+                         const BlockLogicalCoordHolder<Dim>& holder) {
+  os << "BlockLogicalCoordHolder with blockid " << holder.block_id
+     << " and coords " << holder.block_logical_coords;
+  return os;
+}
+
+template <size_t Dim, typename Frame>
+std::vector<BlockLogicalCoordHolder<Dim>> block_logical_coordinates(
+    const Domain<Dim, Frame>& domain,
+    const tnsr::I<DataVector, Dim, Frame>& x) noexcept {
+  const size_t num_pts = get<0>(x).size();
+  std::vector<BlockLogicalCoordHolder<Dim>> block_coord_holders(num_pts);
+  std::vector<tnsr::I<double, Dim, Frame>> points_with_no_block;
+  for (size_t s = 0; s < num_pts; ++s) {
+    tnsr::I<double, Dim, Frame> x_frame(0.0);
+    for (size_t d = 0; d < Dim; ++d) {
+      x_frame.get(d) = x.get(d)[s];
+    }
+    auto& x_logical = block_coord_holders[s].block_logical_coords;
+    // Check which block this point is in. Each point will be in one
+    // and only one block, unless it is on a shared boundary.  In that
+    // case, choose the block with the smallest block_id.
+    bool found_block = false;
+    for (const auto& block : domain.blocks()) {
+      x_logical = block.coordinate_map().inverse(x_frame);
+      bool is_contained = true;
+      for (size_t d = 0; d < Dim; ++d) {
+        // Assumes that logical coordinates go from -1 to +1 in each
+        // dimension.
+        if (x_logical.get(d) < -1.0 or x_logical.get(d) > 1.0) {
+          // Point is not in this block.  Check the next block.
+          is_contained = false;
+          break;
+        }
+      }
+      if (is_contained) {
+        // Point is in this block.  Don't bother checking subsequent
+        // blocks.
+        block_coord_holders[s].block_id = block.id();
+        found_block = true;
+        break;
+      }
+    }
+    if (not found_block) {
+      points_with_no_block.emplace_back(std::move(x_frame));
+    }
+  }
+  if (not points_with_no_block.empty()) {
+    ERROR("Found points that are not in any block\n: x_frame = "
+          << points_with_no_block);
+  }
+  return block_coord_holders;
+}
+
+// Explicit instantiations
+/// \cond
+template struct BlockLogicalCoordHolder<1>;
```

Comment:
using GENERATE_INSTANTIATIONS looks like it would help. Maybe even in the below one using `BOOST_PP_IF()`?